collector/perf_linux.go

```diff
@@ -0,0 +1,604 @@
+			prometheus.BuildFQName(
+				namespace,
+				perfSubsystem,
+				"cache_ll_read_hit_total",
+			),
+			"Number last level read hits",
+			[]string{"cpu"},
+			nil,
+		),
+		"LastLevelReadMiss": prometheus.NewDesc(
+			prometheus.BuildFQName(
+				namespace,
+				perfSubsystem,
+				"cache_ll_read_misses_total",
+			),
+			"Number last level read misses",
+			[]string{"cpu"},
+			nil,
+		),
+		"LastLevelWriteHit": prometheus.NewDesc(
+			prometheus.BuildFQName(
+				namespace,
+				perfSubsystem,
+				"cache_ll_write_hit_total",
+			),
+			"Number last level write hits",
+			[]string{"cpu"},
+			nil,
+		),
+		"LastLevelWriteMiss": prometheus.NewDesc(
+			prometheus.BuildFQName(
+				namespace,
+				perfSubsystem,
+				"cache_ll_write_miss_total",
+			),
+			"Number last level write misses",
+			[]string{"cpu"},
+			nil,
+		),
+		"BPUReadHit": prometheus.NewDesc(
+			prometheus.BuildFQName(
+				namespace,
+				perfSubsystem,
+				"cache_bpu_read_hit_total",
+			),
+			"Number BPU read hits",
+			[]string{"cpu"},
+			nil,
+		),
+		"BPUReadMiss": prometheus.NewDesc(
+			prometheus.BuildFQName(
+				namespace,
+				perfSubsystem,
+				"cache_bpu_read_miss_total",
+			),
+			"Number BPU read misses",
+			[]string{"cpu"},
+			nil,
+		),
+	}
+)
+
+func init() {
+	registerCollector(perfSubsystem, defaultDisabled, NewPerfCollector)
+}
+
+// perfCollector is a Collecter that uses the perf subsystem to collect
+// metrics. It uses perf_event_open an ioctls for profiling. Due to the fact
+// that the perf subsystem is highly dependent on kernel configuration and
+// settings not all profiler values may be exposed on the target system at any
+// given time.
+type perfCollector struct{}
+
+// initializePerfProfilers is used to setup the global map of profilers. For
+// each type of profiler it creates one profiler per CPU.
+func initializePerfProfilers() error {
+	ncpus := runtime.NumCPU()
+	for i := 0; i < ncpus; i++ {
+		// Use -1 to profile all processes on the CPU, see:
+		// man perf_event_open
+		perfHwProfilers[i] = perf.NewHardwareProfiler(-1, i)
+		if err := perfHwProfilers[i].Start(); err != nil {
+			log.Errorf(
+				"Failed to start perf hardware profiler on CPU (%d), %v: %+v",
+				i,
+				perfCheckMsg,
+				err,
+			)
+			return err
+		}
+		perfSwProfilers[i] = perf.NewSoftwareProfiler(-1, i)
+		if err := perfSwProfilers[i].Start(); err != nil {
+			log.Errorf(
+				"Failed to start perf software profiler on CPU (%d), %v: %+v",
+				i,
+				perfCheckMsg,
+				err)
+			return err
+		}
+		perfCacheProfilers[i] = perf.NewCacheProfiler(-1, i)
+		if err := perfCacheProfilers[i].Start(); err != nil {
+			log.Errorf(
+				"Failed to start perf cache profiler on CPU (%d), %v: %+v",
+				i,
+				perfCheckMsg,
+				err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// NewPerfCollector returns a new perf based collector, it creates a profiler
+// per CPU.
+func NewPerfCollector() (Collector, error) {
+	var err error
+	perfInitializeOnce.Do(func() {
```

Comment:
Still not convinced we need this. If you make the variables members of the perfCollector struct, it will be called only once. And even if it gets called twice by accident, I'd assume it would fail which is the right thing to do.

Reply:
I think you are right, I dug into the collector code a bit more and it seems unneeded.